doc/neps/nep-0049.rst

```diff
@@ -0,0 +1,334 @@
+
+NumPy C-API functions
+=====================
+
+.. c:type:: PyDataMem_Handler
+
+    A struct to hold function pointers used to manipulate memory
+
+    .. code-block:: c
+
+        typedef struct {
+            char name[128];  /* multiple of 64 to keep the struct aligned */
+            PyDataMem_AllocFunc *alloc;
+            PyDataMem_ZeroedAllocFunc *zeroed_alloc;
+            PyDataMem_FreeFunc *free;
+            PyDataMem_ReallocFunc *realloc;
+        } PyDataMem_Handler;
+
+    where the function's signatures are
+
+    .. code-block:: c
+
+        typedef void *(PyDataMem_AllocFunc)(size_t size);
+        typedef void *(PyDataMem_ZeroedAllocFunc)(size_t nelems, size_t elsize);
+        typedef void (PyDataMem_FreeFunc)(void *ptr, size_t size);
+        typedef void *(PyDataMem_ReallocFunc)(void *ptr, size_t size);
+
+.. c:function:: const PyDataMem_Handler * PyDataMem_SetHandler(PyDataMem_Handler *handler)
+
+   Sets a new allocation policy. If the input value is ``NULL``, will reset
+   the policy to the default. Returns the previous policy, ``NULL`` if the
+   previous policy was the default. We wrap the user-provided functions
+   so they will still call the Python and NumPy memory management callback
+   hooks. All the function pointers must be filled in, ``NULL`` is not
+   accepted.
+
+.. c:function:: const char * PyDataMem_GetHandlerName(PyArrayObject *obj)
+
+   Return the const char name of the ``PyDataMem_Handler`` used by the
+   ``PyArrayObject``. If ``NULL``, return the name of the current global policy
+   that will be used to allocate data for the next ``PyArrayObject``.
+
+
+Sample code
+===========
+
+This code adds a 64-byte header to each ``data`` pointer and stores information
+about the allocation in the header. Before calling ``free``, a check ensures
+the ``sz`` argument is correct.
+
+.. code-block:: c
+
+    #define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+    #include <numpy/arrayobject.h>
+    NPY_NO_EXPORT void *
+
+    shift_alloc(size_t sz) {
+        char *real = (char *)malloc(sz + 64);
+        if (real == NULL) {
+            return NULL;
+        }
+        snprintf(real, 64, "originally allocated %ld", (unsigned long)sz);
+        return (void *)(real + 64);
+    }
+
+    NPY_NO_EXPORT void *
+    shift_zero(size_t sz, size_t cnt) {
+        char *real = (char *)calloc(sz + 64, cnt);
+        if (real == NULL) {
+            return NULL;
+        }
+        snprintf(real, 64, "originally allocated %ld via zero",
+                 (unsigned long)sz);
+        return (void *)(real + 64);
+    }
+
+    NPY_NO_EXPORT void
+    shift_free(void * p, npy_uintp sz) {
+        if (p == NULL) {
+            return ;
+        }
+        char *real = (char *)p - 64;
+        if (strncmp(real, "originally allocated", 20) != 0) {
+            fprintf(stdout, "uh-oh, unmatched shift_free, "
+                    "no appropriate prefix\\n");
+            /* Make gcc crash by calling free on the wrong address */
```

Comment:
I assume by gcc you mean "the c runtime"?

Reply:
fixed